src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +117,388 @@ public class ConsensusManager
                 utxoHash = await this.consensusRules.RewindAsync().ConfigureAwait(false);
             }
 
-            this.concurrentChain.SetTip(this.Tip);
+            lock (this.treeLock)
+            {
+                this.chainedHeaderTree.Initialize(this.Tip, !blockStoreDisabled);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A list of headers are presented from a given peer,
+        /// we'll attempt to connect the headers to the tree and if new headers are found they will be queued for download.
+        /// </summary>
+        /// <param name="peerId">The peer that is providing the headers.</param>
+        /// <param name="headers">The list of new headers.</param>
+        /// <returns>The last chained header that is connected to the tree.</returns>
+        public ChainedHeader HeadersPresented(int peerId, List<BlockHeader> headers)
+        {
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(peerId), peerId, nameof(headers), nameof(headers.Count), headers.Count);
+
+            ConnectNewHeadersResult newHeaders = null;
+
+            lock (this.treeLock)
+            {
+                newHeaders = this.chainedHeaderTree.ConnectNewHeaders(peerId, headers);
+                this.blockPuller.NewTipClaimed(peerId, newHeaders.Consumed); // TODO: should this be inside the lock?
+            }
+
+            if (newHeaders?.DownloadTo != null)
+            {
+                this.DownloadBlocks(newHeaders.ToHashList(), this.ProcessDownloadedBlock);
+            }
+
+            this.logger.LogTrace("(-):'{0}'", newHeaders?.Consumed);
+            return newHeaders?.Consumed;
+        }
+
+        /// <summary>
+        /// A peer was disconnected, clean all its claimed headers that are above our consensus tip.
```

Comment:
On this level, we rather just inform the depending components and only one of those components is actually cleaning the headers, but that is its internal thing I guess